### PR TITLE
Migration of zuliprc location to a config directory

### PR DESCRIPTION
**Pull request by Niloth-p**
### What does this PR do, and why?

Adds the following features:

- Support recognizing zuliprc from multiple select locations. 
  - `~/zuliprc`
  - the user's downloads directory
  - the user's config directory

- Move zuliprc creation into user's config directory.
The new path would look like `.config/zulip-terminal/zuliprc`.

- Create realm-wise zuliprc files in the configuration directory.
Each realm gets its own directory.
The new directory structure would look like:
```
<user's config directory>/zulip-terminal/
  org1-name/zuliprc
  org2-name/zuliprc
  zuliprc
```

- Add new positional argument - prefix of realm name to get its configuration.
New command examples:
`zulip-term zuli`
`zulip-term "full organization name"`

- Add -o argument to list realms with stored zuliprc.
```
$zulip-term -o
The following organizations are available:
  Zulip Community
  Another Organization
```

- Add -n argument to login to a realm, and create config.
```
$zulip-term -n
Please enter your credentials to login into your Zulip organization.
```

<details>
<summary>What does this PR do, and why? (expired)</summary>

### What does this PR do, and why?
Moves config location to 'XDG_CONFIG/zulip/'.
Makes zuliprc files dot files - .zuliprc.
Use independent org.zuliprc files for each organization.

Does not yet separate the zterm config from the api config.
Does not yet update the documents (README.md, FAQ.md, etc).
New tests not added.

The code is not yet meant for review.
This is only a draft to play around with the configuration in order to finalize what behavior is desired.
</details>

### External discussion & connections
- [x] Discussed in **#zulip-terminal** in `Moving config`
- [ ] Fully fixes #
- [x] Partially fixes issue #678
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
- [x] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### CODE REVIEW

File: zulipterminal/cli/run.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ class SettingData(NamedTuple):
 
 DOWNLOADED_PATH_ZULIPRC = str(downloads_file_path() / "zuliprc")
 HOME_PATH_ZULIPRC = str(Path.home() / "zuliprc")
-CONFIG_PATH_ZULIPRC = str(config_file_path() / "zulip-terminal" / "zuliprc")
```
Comment (Niloth-p) [inline zulipterminal/cli/run.py]: Changed names here.
Felt CONFIG_PATH_ZULIPRC was a more consistent name for the single zuliprc in config.
Feel ZULIP_CONFIG_PATH would be better when pointing to the directory. Could use CONFIG_PATH_ZULIP too.
Would be open to rebasing any name changes if the noise of this change is more important than the consistency of the name in the initial commits.

File: zulipterminal/cli/run.py [hunk 2]
```diff
@@ -343,13 +349,19 @@ def _write_zuliprc(
...
         DOWNLOADED_PATH_ZULIPRC,
         HOME_PATH_ZULIPRC,
     ]
     valid_locations = [
         location for location in zuliprc_locations if Path(location).exists()
     ]
+    valid_locations.extend(
+        str(file) for file in Path(ZULIP_CONFIG_PATH).glob("zuliprc")
+    )
+    valid_locations.extend(
```
Comment (Niloth-p) [inline zulipterminal/cli/run.py:385]: The current implementation explicitly does not use `Path(ZULIP_CONFIG_PATH).glob("**/zuliprc")`, to avoid the situation (in later commits) where the realm has multiple zuliprc files inside it (in folders or as loose file). This implementation will only recognize the outermost zuliprc within a realm's directory, irrespective of the subdirectories' contents.

File: zulipterminal/cli/run.py [hunk 3]
```diff
@@ -448,6 +448,30 @@ def list_themes() -> str:
...
+        matching_dirs_string = "\n".join(str(dir) for dir in matching_dirs)
+        exit_with_error(
+            f"Found multiple organizations starting with '{realm_name_prefix}':\n"
+            f"{matching_dirs_string}\n"
+        )
+    elif len(matching_dirs) == 0:
+        exit_with_error(
+            f"Could not find any organizations starting with '{realm_name_prefix}'"
+        )
+    matching_files = list(matching_dirs[0].glob("zuliprc"))
```
Comment (Niloth-p) [inline zulipterminal/cli/run.py:509]: If we used path.glob(**/zuliprc) in the previous commit, this is where we would need to add another case where len(matching_files) > 1.

File: zulipterminal/cli/run.py [hunk 4]
```diff
@@ -448,6 +448,30 @@ def list_themes() -> str:
     )
 
 
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
+    if len(realm_name_prefix) < 3:
```
Comment (Niloth-p) [inline zulipterminal/cli/run.py:492]: Assumption: There are no zulip realms whose name property contains 2 or fewer characters.
We could remove this case.

File: zulipterminal/cli/run.py [hunk 5]
```diff
@@ -387,15 +398,21 @@ def check_for_default_zuliprc() -> str:
...
+            if zuliprc_str == ""
+            else path.expanduser(zuliprc_str)
+        )
+        if not new_realm
+        else ""
     )
     while zuliprc_path == "" or not path.exists(zuliprc_path):
         try:
-            zuliprc_path = fetch_zuliprc(zuliprc_path)
+            zuliprc_path = fetch_zuliprc(zuliprc_path, new_realm)
```
Comment (Niloth-p) [inline zulipterminal/cli/run.py:415]: Re. naming,
used 'realm' for all internal references,
and 'org' / 'organization' for all references visible to the user.

**Review by Niloth-p — COMMENTED** [review]
Updated the PR description, to describe the new features.
Each item in that feature list is its own commit (in the same order), with preparatory commits in between (mentioned for easier reviewing).

I have not yet added tests in this branch.
The ones I had were too tightly tied to the functions that were introduced.
Planning to re-write them to directly test main([]) instead.

#### Reference: Testing cases:
Feel free to come up with any missing / edge cases.
- default call (without positional argument)
	- loose zuliprc file in config
	- zuliprc files in config inside realms directories
		- one realm
		- multiple realms
	- mix of both loose zuliprc and inside realms in config
	- several zuliprc files found error
	- no zuliprc in any locations
- positional argument
	- too few characters (<3)
	- no org found
	- several matching orgs
	- one matching org, but zuliprc is inside subdirectory
	- one matching org
	- full name instead of prefix
	- use along with --config-file option & -c options
- login newly into a few servers, test the created directory structure
- n option:
	- if configuration for that realm already exists
	- if a directory with that realm name exists, but it does not have a zuliprc file inside it
- o option:
	- no orgs found
	- several orgs found

File: zulipterminal/cli/run.py [hunk 6]
```diff
@@ -322,11 +371,48 @@ def _write_zuliprc(
...
+    count = len(valid_locations)
+    if count > 1:
+        valid_locations_string = "\n".join(valid_locations)
+        exit_with_error(
+            f"Found multiple zuliprc files at:\n{valid_locations_string}\n"
+            "Please retry by specifying the path to your target zuliprc file."
+        )
+    elif count == 1:
+        return valid_locations[0]
+    return ""
```
Comment (rsashank) [inline zulipterminal/cli/run.py:398]: This implementation seems great! However, should we consider giving priority to `HOME_PATH_ZULIPRC` if multiple configuration files are found, since that was previously the location of zuliprc?

If users want to use a different configuration file instead of the default, they can specify a specific target zuliprc file to override the default. We can notify the config file used while loading ZT, similar to how we do with the settings.

Alternatively, instead of exiting with an error, we could include an index and prompt the user to choose which organization they would like to log into.

For example:

```shell
Choose which organization you'd like to log in to (Enter the number):
1: Zulip Community: chat.zulip.org
2: CTF Chat: ctf.zulipchat.com
3: Example: example.zulip.com
```

File: zulipterminal/cli/run.py [hunk 7]
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
...
+
+
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
+    if len(realm_name_prefix) < 3:
+        exit_with_error("Organization name prefix must be at least 3 characters long")
+    matching_dirs = [
+        dir
+        for dir in Path(ZULIP_CONFIG_PATH).glob("*/")
+        if dir.name.startswith(realm_name_prefix)
+    ]
```
Comment (rsashank) [inline zulipterminal/cli/run.py:498]: Can we make this case sensitive?

Maybe:

```python
    matching_dirs = [
        dir
        for dir in Path(ZULIP_CONFIG_PATH).glob("*/")
        if dir.name.startswith(realm_name_prefix)
        or dir.name.lower().startswith(realm_name_prefix)
    ]
```

File: zulipterminal/cli/run.py [hunk 8]
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
...
+    if len(matching_dirs) > 1:
+        matching_dirs_string = "\n".join(str(dir) for dir in matching_dirs)
+        exit_with_error(
+            f"Found multiple organizations starting with '{realm_name_prefix}':\n"
+            f"{matching_dirs_string}\n"
+        )
```
Comment (rsashank) [inline zulipterminal/cli/run.py:504]: I'm not sure how feasible this is, but should we consider prompting the user to select which directory to log into from the matched directories?

File: zulipterminal/cli/run.py [hunk 9]
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
...
+    elif len(matching_dirs) == 0:
+        exit_with_error(
+            f"Could not find any organizations starting with '{realm_name_prefix}'"
+        )
```
Comment (rsashank) [inline zulipterminal/cli/run.py:508]: Maybe we could suggest using zulip-term -o to show the list of available organizations. Alternatively, we could display the list along with the error message?

**Review by rsashank — COMMENTED** [review]
Great work, @Niloth-p! I manually tested it, and it works well. I added a few comments in some areas where I think it would be nice to consider changes, but those are more opinion-based.

This looks very useful!

**Comment by rsashank** [comment]
Thanks for updating this, @Niloth-p! I'll leave it for the mentor and maintainers to review. I'm removing the buddy review label now.

File: zulipterminal/cli/run.py [hunk 10]
```diff
@@ -452,6 +452,30 @@ def list_themes() -> str:
     )
 
 
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
```
Comment (neiljp) [inline zulipterminal/cli/run.py:496]: I'll wait to see this updated for the new version first, but I expect this and later commits would be the mapping of aliases to zuliprcs instead.

File: zulipterminal/cli/run.py [hunk 11]
```diff
@@ -305,8 +311,13 @@ def fetch_zuliprc(zuliprc_path: str) -> str:
...
         else "any of the following locations:\n  " + "\n  ".join(supported_locations)
     )
+    missing_zuliprc_text = (
+        ""
+        if new_realm
+        else f"{in_color('red', f'zuliprc file was not found at {locations_checked}')}"
+    )
     print(
-        f"{in_color('red', f'zuliprc file was not found at {locations_checked}')}"
+        f"{missing_zuliprc_text}"
```
Comment (neiljp) [inline zulipterminal/cli/run.py:320]: This will likely be simplified with the other changes I suggested?

File: zulipterminal/cli/run.py [hunk 12]
```diff
@@ -157,6 +157,12 @@ def parse_args(argv: List[str]) -> argparse.Namespace:
         action="store_true",
         help="list all the organizations that you have configurations for, and exit",
     )
+    parser.add_argument(
+        "-n",
+        "--new-organization",
+        action="store_true",
+        help="login to a new organization",
+    )
```
Comment (neiljp) [inline zulipterminal/cli/run.py:165]: It doesn't quite fit within the remit of the zuliprc movement work, but it'd be good to have `-c` work with creating a zuliprc file from username/password, which I don't think it does right now?

**Review by neiljp — COMMENTED** [review]
@Niloth-p I hope this review helps get you back on track with this work :+1:

The discussion in the channel seems to have helped too :)
